### PR TITLE
Implementation of Read for Python

### PR DESCRIPTION
**Pull request by swgillespie**
This commit implements read_resource functionality for Python in a
manner identical to the NodeJS implementation. If an "id" option is
passed to a resource via ResourceOptions on construction, that resource
will be read and not created.

### CODE REVIEW

**Comment by swgillespie** [comment]
StackReference is a short hop away, once this PR merges: https://github.com/pulumi/pulumi/compare/swgillespie/stack-reference?expand=1

File: sdk/python/lib/pulumi/resource.py [hunk 1]
```diff
@@ -79,7 +79,12 @@ class ResourceOptions:
     to mark certain ouputs as a secrets on a per resource basis.
     """
 
+    id: Optional[str]
+    """
+    An optional exisitng ID to load, rather than create.
```
Comment (pgavlin) [inline sdk/python/lib/pulumi/resource.py]: `s/exisitng/existing/`

File: sdk/python/lib/pulumi/runtime/resource.py [hunk 2]
```diff
@@ -118,6 +118,118 @@ class ResourceResolverOperations(NamedTuple):
...
+    # Prepare the resource, similar to a RegisterResource. Reads are deliberately similar to RegisterResource except
+    # that we are populating the Resource object with properties associated with an already-live resource.
+    #
+    # Same as below, we initialize the URN property on the resource, which will always be resolved.
+    log.debug(f"preparing read resource for RPC")
+    urn_future = asyncio.Future()
+    urn_known = asyncio.Future()
+    urn_secret = asyncio.Future()
+    urn_known.set_result(True)
+    urn_secret.set_result(True)
```
Comment (pgavlin) [inline sdk/python/lib/pulumi/runtime/resource.py]: Does this imply that the URN will be treated as a secret?

File: sdk/python/lib/pulumi/runtime/resource.py [hunk 3]
```diff
@@ -118,6 +118,118 @@ class ResourceResolverOperations(NamedTuple):
...
+                    # If no monitor is available, we'll need to fake up a response, for testing.
+                    return RegisterResponse(create_test_urn(ty, name), None, resolver.serialized_props)
+
+            resp = await asyncio.get_event_loop().run_in_executor(None, do_rpc_call)
+
+        except Exception as exn:
+            log.debug(f"exception when preparing or executing rpc: {traceback.format_exc()}")
+            rpc.resolve_outputs_due_to_exception(resolvers, exn)
+            resolve_urn_exn(exn)
+            if resolve_id is not None:
```
Comment (pgavlin) [inline sdk/python/lib/pulumi/runtime/resource.py]: This `if` shouldn't be necessary: `resolve_id` will never be `None`

File: sdk/python/lib/pulumi/runtime/resource.py [hunk 4]
```diff
@@ -118,6 +118,118 @@ class ResourceResolverOperations(NamedTuple):
...
+                resolve_id(None, False, exn)
+            raise
+
+        log.debug(f"resource read successful: ty={ty}, urn={resp.urn}")
+        resolve_urn(resp.urn)
+        resolve_id(resolve_id, True, None) # Read IDs are always known.
```
Comment (pgavlin) [inline sdk/python/lib/pulumi/runtime/resource.py]: The first parameter here should be `resp.id`.

**Pull request merged**